### PR TITLE
TRUNK-5934:Collapsible "if" statements should be merged

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/TRUNK-5934

cc @dkayiwa  @sacull

### CODE REVIEW

**Comment by dkayiwa** [comment]
@sherrif10 this category of tickets is no longer fit for you. It is too simple for you to learn from it.

**Comment by sherrif10** [comment]
@dkayiwa true dat, i can un assign myself for someone else

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java [hunk 1]
```diff
@@ -192,12 +192,12 @@ String getPostDataTypeString(Database database) {
 	String getDefaultClause(ColumnConfig column, Database database) {
 		String clause = "";
 		String defaultValue = column.getDefaultValue();
-		if (defaultValue != null) {
-			if (database instanceof MySQLDatabase) {
+		//collapsible if stamements should be merged
+		if (defaultValue != null && database instanceof MySQLDatabase){
+//			if (database instanceof MySQLDatabase) {
```
Comment (sacull) [inline api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:197]: Don't leave such comments. They cause unnecessary garbage in the code.

File: api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java [hunk 2]
```diff
@@ -756,15 +756,14 @@ public void remove() {
 				
 				if (prevIndex == NONE) {
 					throw new IllegalStateException();
-				} else if (prevIndex >= 0) {
-					if (elements[prevIndex] == prevItem) {
+				} else if ((prevIndex >= 0) && (elements[prevIndex] == prevItem)) {
+//					if (elements[prevIndex] == prevItem) {
```
Comment (sacull) [inline api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java:760]: Same as above.

File: api/src/test/java/liquibase/ext/sqlgenerator/core/ModifyColumnGeneratorTest.java [hunk 3]
```diff
@@ -52,7 +52,7 @@ public void shouldGenerateSql() {
 		config.setType("int");
 		
 		ModifyColumnStatement statement = new ModifyColumnStatement("some_schema", "some_table", config);
-		Sql[] sql = generator.generateSql(statement, new MySQLDatabase(), null);
+		Sql[] sql = generator.generateSql(statement, new MySQLDatabase(),null);
```
Comment (sacull) [inline api/src/test/java/liquibase/ext/sqlgenerator/core/ModifyColumnGeneratorTest.java:55]: And why did you remove the space here? : P

**Review by sacull — CHANGES_REQUESTED** [review]

**Comment by sacull** [comment]
@sherrif10 Since you have already started, you may finish, especially since, as you can see, you still have to improve a bit. ;)

After that, we will try to make sure that the intro tickets will be enough for everyone.

**Comment by sherrif10** [comment]
@sacull no problem with that please, let me collect that small mistakes

**Comment by sacull** [comment]
@sherrif10 We all learn all the time. Have a nice studying. ;)

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 4]
```diff
@@ -125,12 +125,14 @@ public void validate(Object obj, Errors errors) {
 				}
 			}
 			
+			
 			if (!StringUtils.isBlank(user.getEmail())) {
```
Comment (sherrif10) [inline api/src/main/java/org/openmrs/validator/UserValidator.java:129]: @sacull  @dkayiwa  so i think you followed this commit, me trying to change the structure of these if statements line 129 breaks the testing in UserValidatorTest. My first approach on this was TODO  like
  **if((!StringUtils.isBlank(user.getEmails()) ||(!isEmailValid(user.getEmail())) ){
}**
but that lead to breakage of testing, i didnt wanted to go away from refactoring from this

Comment (sacull) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: Why || not &&?

Comment (sacull) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: Both conditions must be met, not just one of them. ;)

Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: That method using && block tests let me use asap and you have a look

Comment (sacull) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: I already know where the problem is, you have too many parentheses and just one of you moved at the end instead of being in the middle.

Take a look at this entry and consider why it will work and how it is different from yours: 
`if (!StringUtils.isBlank(user.getEmail()) && !isEmailValid(user.getEmail())) {}`

Comment (sacull) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: One more thing, if you want to continue this PR, remember to assign a JIRA ticket to yourself.

Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: True dat @sacull Actually with that matter we need to add an else close to get rid of NPE if we just used it like that we would fall in NPE  which tends to happen in UserValidatorTest. take a look of how it looks now

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 5]
```diff
@@ -126,17 +126,20 @@ public void validate(Object obj, Errors errors) {
...
 			
-			if (!StringUtils.isBlank(user.getEmail())) {
-				if(!isEmailValid(user.getEmail())) {
+			if (!StringUtils.isBlank(user.getEmail()) && !isEmailValid(user.getEmail())) {
+
 					errors.rejectValue("email", "error.email.invalid");
 				}
-			}
 			
+			 else {
```
Comment (sacull) [inline api/src/main/java/org/openmrs/validator/UserValidator.java:134]: Try to stick to the rule that "else" is on the same line as the earlier brace.

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 6]
```diff
@@ -126,17 +126,20 @@ public void validate(Object obj, Errors errors) {
...
 			
 			ValidateUtil.validateFieldLengths(errors, obj.getClass(), "username", "systemId", "retireReason");
+			 }
```
Comment (sacull) [inline api/src/main/java/org/openmrs/validator/UserValidator.java:137]: Space before brace?

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 7]
```diff
@@ -126,17 +126,20 @@ public void validate(Object obj, Errors errors) {
 			}
 			
 			
-			if (!StringUtils.isBlank(user.getEmail())) {
-				if(!isEmailValid(user.getEmail())) {
+			if (!StringUtils.isBlank(user.getEmail()) && !isEmailValid(user.getEmail())) {
+
```
Comment (sacull) [inline api/src/main/java/org/openmrs/validator/UserValidator.java:130]: Try not to leave such blank lines, especially if there is only one command in the braces.

**Review by sacull — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 6 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:136]: Finished

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 7 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: Finished that

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 5 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java]: Finished thanks

File: api/src/main/java/org/openmrs/validator/UserValidator.java (hunk 7 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: Anything else @sacull

Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:129]: @dkayiwa  feel free to merge this ka PR  though i bumped into it and we close it

**Comment by dkayiwa** [comment]
Will wait to hear what @sacull says about it. 😊

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java [hunk 8]
```diff
@@ -192,12 +192,10 @@ String getPostDataTypeString(Database database) {
 	String getDefaultClause(ColumnConfig column, Database database) {
 		String clause = "";
 		String defaultValue = column.getDefaultValue();
-		if (defaultValue != null) {
-			if (database instanceof MySQLDatabase) {
+		if (defaultValue != null && database instanceof MySQLDatabase){
```
Comment (sacull) [inline api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:195]: There shouldn't be a space before the brace here?

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java [hunk 9]
```diff
@@ -192,12 +192,10 @@ String getPostDataTypeString(Database database) {
...
 				clause += " DEFAULT " + DataTypeFactory.getInstance().fromObject(defaultValue, database)
 					.objectToSql(defaultValue, database);
```
Comment (sacull) [inline api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:197]: Check the formatting here, there seems to be one unnecessary space at the beginning.

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java [hunk 10]
```diff
@@ -192,12 +192,10 @@ String getPostDataTypeString(Database database) {
...
 			}
```
Comment (sacull) [inline api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:198]: Here, perhaps too much by one tab.

File: api/src/main/java/org/openmrs/util/MemoryAppender.java [hunk 11]
```diff
@@ -76,16 +76,15 @@ protected static MemoryAppender createAppender(
 		MemoryAppender appender = null;
 		if (APPENDERS.containsKey(name)) {
 			appender = APPENDERS.get(name).get();
-
-			if (appender != null) {
-				if (appender.bufferSize != theBufferSize) {
-					ThreadSafeCircularFifoQueue<LogEvent> oldBuffer = appender.buffer;
-					appender.buffer = new ThreadSafeCircularFifoQueue<>(theBufferSize);
-					appender.bufferSize = theBufferSize;
+            //we can use && operator for conditions when  both either  true or false
```
Comment (sacull) [inline api/src/main/java/org/openmrs/util/MemoryAppender.java:79]: Such comments are not really needed in the code.

File: api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java [hunk 12]
```diff
@@ -756,15 +756,13 @@ public void remove() {
 				
 				if (prevIndex == NONE) {
 					throw new IllegalStateException();
-				} else if (prevIndex >= 0) {
-					if (elements[prevIndex] == prevItem) {
+				} else if ((prevIndex >= 0) && (elements[prevIndex] == prevItem)) {
 						internalRemoveAtIndex(prevIndex);
```
Comment (sacull) [inline api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java:761]: Empty line?

File: api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java [hunk 13]
```diff
@@ -756,15 +756,13 @@ public void remove() {
...
 
-						if (prevIndex != read) {
+			      	}	else if (prevIndex != read) {
```
Comment (sacull) [inline api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java:762]: Tab before tab?

Comment (sacull) [reply api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java:762]: Additionally, there are spaces before the first brace?

**Review by sacull — CHANGES_REQUESTED** [review]

**Comment by sacull** [comment]
@sherrif10 I'm really sorry, but you left a lot of minor bugs in the code, mostly related to tabs and spaces, so the formatting of the code itself. It works, but we also want it to be of good quality?

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java (hunk 9 above)
Comment (sherrif10) [reply api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:197]: Even here no space except if am not seeing well , that code is in the same line as before

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java (hunk 10 above)
Comment (sherrif10) [reply api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:198]: Yes its should be like that, do you want to introduce a new braces

File: api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java (hunk 13 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java:762]: Finished all check again, and left some comments

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java (hunk 9 above)
Comment (sacull) [reply api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:197]: Line 195 - 2 tabs, line 196 - 4 tabs, right? Isn't it better if there were 3 instead of 4? Unless GitHub is distorting and showing something different than what you see in the IDE.

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java (hunk 8 above)
Comment (sherrif10) [reply api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:195]: There is no space except if you are saying i add space

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java (hunk 10 above)
Comment (sacull) [reply api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:198]: No extra curly braces. The point is that there are two tabs before "if" and three tabs before the brace associated with this "if". Or GitHub shows something different than what you see in your IDE.

**Comment by sherrif10** [comment]
Any more identation @sacull ?

**Comment by sacull** [comment]
@sherrif10 I will clone your branch and check if GitHub shows what is for real or something else and make the mistake of pointing you to places that are not there.

File: api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java (hunk 9 above)
Comment (sacull) [reply api/src/main/java/liquibase/ext/sqlgenerator/core/ModifyColumnGenerator.java:197]: image

File: api/src/main/java/org/openmrs/util/MemoryAppender.java [hunk 14]
```diff
@@ -76,16 +76,14 @@ protected static MemoryAppender createAppender(
...
-					appender.buffer = new ThreadSafeCircularFifoQueue<>(theBufferSize);
-					appender.bufferSize = theBufferSize;
+			if (appender != null && appender.bufferSize != theBufferSize) {
+				ThreadSafeCircularFifoQueue<LogEvent> oldBuffer = appender.buffer;
+				appender.buffer = new ThreadSafeCircularFifoQueue<>(theBufferSize);
+				appender.bufferSize = theBufferSize;
+				appender.buffer.addAll(oldBuffer);
 					appender.buffer.addAll(oldBuffer);
 				}
 			}
```
Comment (sacull) [inline api/src/main/java/org/openmrs/util/MemoryAppender.java:86]: image

File: api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java [hunk 15]
```diff
@@ -756,15 +756,12 @@ public void remove() {
...
 						internalRemoveAtIndex(prevIndex);
-
-						if (prevIndex != read) {
+			      	} else if (prevIndex != read) {
 							nextIndex = Math.max(prevIndex, read);
-						}
-					}
-				}
+		    	}
+
```
Comment (sacull) [inline api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java:764]: image

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 16]
```diff
@@ -125,13 +125,13 @@ public void validate(Object obj, Errors errors) {
...
 					errors.rejectValue("email", "error.email.invalid");
-				}
-			}
-			
+				} else {
+
```
Comment (sacull) [inline api/src/main/java/org/openmrs/validator/UserValidator.java:132]: image

**Review by sacull — CHANGES_REQUESTED** [review]

**Comment by sacull** [comment]
@sherrif10 I'm really not doing this on purpose, sorry.

**Comment by sherrif10** [comment]
interesting @sacull

File: api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java (hunk 15 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/util/ThreadSafeCircularFifoQueue.java:764]: Have you tried and see how tha full code ends up with whole code format. To me reverting that lines will make the code to look wierd

File: api/src/main/java/org/openmrs/util/MemoryAppender.java (hunk 14 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/util/MemoryAppender.java:86]: only line 84 should be changed here, finished it

**Comment by sherrif10** [comment]
Do you need more identation @sacull  haaa

**Comment by sacull** [comment]
This is code someone else wrote:
image
image
image


This is code written by you:
image
image
image
image

Lines of the same colour should be in the same columns. You really don't notice it?

Additionally, in the last example:
- there is one tab too many on line 129, really;
- on line 131 - why this empty line, what function does it perform?
- there are spaces on line 133 and there should be tabs, check it out, don't take my word for it.

**Comment by sherrif10** [comment]
Let me know if you are not satisfied with formatting please

**Comment by sacull** [comment]
image

Line 129 - there is still one tab too many.
Line 131 - this line is still unnecessary, why don't you want to delete it, maybe there is some reason I don't know?
Line 132 - there are three tabs here and there should be four.
Line 133 - there are still four spaces here instead of one tab. You may not see it, but it really does matter. (dashes - tab, dots - spaces)

**Comment by sacull** [comment]
I cannot explain it otherwise. Below, what it should look like according to Java convention.

image

**Comment by sherrif10** [comment]
Check again thank you

**Comment by sacull** [comment]
image

Now you've added a tab at the end of line 130, what for?
There were more spaces on lines 131 and 132, this can only be explained by one thing - your IDE inserts four spaces when you hit the tab key, but here I can't help but rather. This is your IDE configuration.

**Comment by sherrif10** [comment]
The problem is with my eclipse i think

**Comment by sacull** [comment]
@sherrif10 It is worth configuring this behavior by the IDE, because in the future you will have to explain it many times why you give spaces instead of tabs.

@dkayiwa After these adventures, please take a look at it, because I must have missed something, I can feel it. ;)

**Comment by dkayiwa** [comment]
@sacull i do not feel confortable merging this pull request if you are not yet satisfied with it.

**Comment by sacull** [comment]
@dkayiwa I found what I missed:
image

@sherrif10 Do you see it in your IDE too?

**Comment by sacull** [comment]
@sherrif10 I can see that your IDE continues to insert spaces for you instead of tabs. Can't be changed?

**Comment by sherrif10** [comment]
That may cause more harm in this code, i will try collectly configure my IDE

**Comment by sacull** [comment]
@sherrif10 This is no longer a decision I can make. @dkayiwa will append this code to the master, so it's up to him. If he writes that spaces can stay then they can stay, if he writes that there must be tabs then you will have to convert them to tabs.

**Comment by dkayiwa** [comment]
@sherrif10 let us do what @sacull recommends.

**Comment by sherrif10** [comment]
@dkayiwa It have already taken time in collecting tabs and staff, @sacull  Are you still not comfortable with tabing

File: api/src/main/java/org/openmrs/validator/UserValidator.java [hunk 17]
```diff
@@ -35,170 +35,173 @@
...
+      log.error("Username Pattern Syntax exception in UserValidator", pex);
+      return false;
+    }
+  }
+
+  /**
+   * Returns true if the given username is a valid e-mail.
+   *
+   * @param username
+   * @return true if valid
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/validator/UserValidator.java:191]: @sherrif10 you seem to have more changes than required by the ticket.

Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:191]: That was there before my new changes

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:191]: Is it required by the ticket?

Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:191]: Yes sure its required since it was formally there

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:191]: Let me put it differently, does the ticket require making all these changes? https://github.com/openmrs/openmrs-core/pull/3531/files

Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:191]: @sacull are you still looking into this. take a look once more

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:191]: @sherrif10 it is best to fist deal with what other reviewer have brought up, before you again ping @sacull 
That way, you do not over burden him.

Comment (sherrif10) [reply api/src/main/java/org/openmrs/validator/UserValidator.java:191]: ok thanks @dkayiwa

**Comment by sacull** [comment]
@sherrif10 Forgive me if you feel bad about my tormenting you with these whitespace. Hope you will keep trying your hand at OpenMRS code.

**Comment by sherrif10** [comment]
No problem its part of learning , am opening a new PR
